File: desktop/WaiComputer.Core/Audio/AudioCompressor.cs
using System.Buffers.Binary;
using Concentus;
using Concentus.Enums;
using Concentus.Oggfile;

namespace WaiComputer.Core.Audio;

/// <summary>Result of compressing a recording before upload.</summary>
public sealed record CompressedAudio(string Path, long ByteCount, double DurationSeconds);

/// <summary>
/// Transcodes the finalized PCM WAV recording to Ogg-Opus before upload. Raw
/// 16 kHz mono PCM is ~110 MB/hour and trips the 200 MB upload ceiling for
/// recordings over ~1h49m; Opus at speech bitrates is ~11 MB/hour. Pure-managed
/// (Concentus), so it runs identically on Windows and Linux.
/// </summary>
public static class AudioCompressor
{
    // Opus only accepts these input sample rates.
    private static readonly int[] SupportedSampleRates = { 8000, 12000, 16000, 24000, 48000 };

    /// <summary>
    /// Transcodes a PCM WAV file to Ogg-Opus, preserving sample rate and channel
    /// count. Streams the source so multi-hour recordings don't load fully into
    /// memory. Overwrites <paramref name="destOpusPath"/> if present.
    /// </summary>
    /// <param name="sourceWavPath">Path to the source PCM WAV file.</param>
    /// <param name="destOpusPath">Path to write the Ogg-Opus file to.</param>
    /// <param name="bitRate">Target Opus bitrate. 24 kbps mono is clean for
    /// speech and keeps a 4-hour meeting near ~45 MB.</param>
    public static CompressedAudio CompressWavToOpus(
        string sourceWavPath,
        string destOpusPath,
        int bitRate = 24_000)
    {
        using var input = new FileStream(sourceWavPath, FileMode.Open, FileAccess.Read);
        var (sampleRate, channels, dataLength) = ReadWavHeader(input);

        if (Array.IndexOf(SupportedSampleRates, (int)sampleRate) < 0)
        {
            throw new NotSupportedException($"Opus does not support {sampleRate} Hz input.");
        }
        if (dataLength <= 0)
        {
            throw new InvalidDataException("WAV contains no PCM data.");
        }

        var encoder = OpusCodecFactory.CreateEncoder((int)sampleRate, channels, OpusApplication.OPUS_APPLICATION_VOIP);
        encoder.Bitrate = bitRate;

        long totalFrames = dataLength / (2L * channels);

        using (var output = new FileStream(destOpusPath, FileMode.Create, FileAccess.Write))
        {
            var oggOut = new OpusOggWriteStream(encoder, output, null, (int)sampleRate);

            const int frameChunk = 16_384;
            var byteBuffer = new byte[frameChunk * channels * 2];
            var sampleBuffer = new short[frameChunk * channels];
            long remaining = dataLength;

            while (remaining > 0)
            {
                int toRead = (int)Math.Min(byteBuffer.Length, remaining);
                int read = input.ReadAtLeast(byteBuffer.AsSpan(0, toRead), toRead, throwOnEndOfStream: false);
                if (read <= 0) break;

                int sampleCount = read / 2;
                for (int i = 0; i < sampleCount; i++)
                {
                    sampleBuffer[i] = BinaryPrimitives.ReadInt16LittleEndian(byteBuffer.AsSpan(i * 2, 2));
                }
                oggOut.WriteSamples(sampleBuffer, 0, sampleCount);
                remaining -= read;
            }

            oggOut.Finish();
        }

        var byteCount = new FileInfo(destOpusPath).Length;
        var durationSeconds = (double)totalFrames / sampleRate;
        return new CompressedAudio(destOpusPath, byteCount, durationSeconds);
    }

    /// <summary>
    /// Walks RIFF chunks to read format + locate PCM data, leaving the stream
    /// positioned at the first PCM byte. Tolerates extra chunks (LIST/fact/etc.).
    /// </summary>
    private static (uint sampleRate, int channels, long dataLength) ReadWavHeader(Stream stream)
    {
        Span<byte> tag = stackalloc byte[4];
        Span<byte> u32 = stackalloc byte[4];

        stream.ReadExactly(tag);
        if (!tag.SequenceEqual("RIFF"u8)) throw new InvalidDataException("Not a RIFF file.");
        stream.ReadExactly(u32); // overall RIFF size (ignored)
        stream.ReadExactly(tag);
        if (!tag.SequenceEqual("WAVE"u8)) throw new InvalidDataException("Not a WAVE file.");

        uint sampleRate = 0;
        ushort channels = 0;
        var haveFormat = false;

        while (true)
        {
            stream.ReadExactly(tag);
            stream.ReadExactly(u32);
            uint chunkSize = BinaryPrimitives.ReadUInt32LittleEndian(u32);

            if (tag.SequenceEqual("fmt "u8))
            {
                Span<byte> fmt = stackalloc byte[16];
                stream.ReadExactly(fmt);
                channels = BinaryPrimitives.ReadUInt16LittleEndian(fmt[2..4]);
                sampleRate = BinaryPrimitives.ReadUInt32LittleEndian(fmt[4..8]);
                haveFormat = true;

                long skip = (long)chunkSize - 16 + (chunkSize & 1); // remainder + padding byte
                if (skip > 0) stream.Seek(skip, SeekOrigin.Current);
            }
            else if (tag.SequenceEqual("data"u8))
            {
                if (!haveFormat) throw new InvalidDataException("WAV 'data' chunk preceded 'fmt '.");
                if (channels < 1) throw new InvalidDataException("WAV has no channels.");
                return (sampleRate, channels, chunkSize);
            }
            else
            {
                stream.Seek(chunkSize + (chunkSize & 1), SeekOrigin.Current);
            }
        }
    }
}

File: android/app/src/main/java/is/waiwai/computer/recording/AudioCompressor.kt
package `is`.waiwai.computer.recording

import android.media.MediaCodec
import android.media.MediaCodecInfo
import android.media.MediaFormat
import android.media.MediaMuxer
import java.io.File
import java.io.RandomAccessFile
import java.nio.ByteBuffer
import java.nio.ByteOrder

/** Result of compressing a recording before upload. */
data class CompressedAudio(val file: File, val byteCount: Long, val durationSeconds: Long)

/** Parsed PCM layout of a WAV file. */
data class WavFormat(
    val sampleRate: Int,
    val channels: Int,
    val dataOffset: Long,
    val dataLength: Long,
)

/**
 * Transcodes the recorded PCM WAV to AAC `.m4a` before upload. Raw 16 kHz mono
 * PCM is ~110 MB/hour and trips the 200 MB upload ceiling for recordings over
 * ~1h49m; AAC at speech bitrates is ~22 MB/hour. AAC encoding via [MediaCodec]
 * is available from API 16, so no minSdk bump is required.
 */
object AudioCompressor {

    private const val TIMEOUT_US = 10_000L
    private const val MAX_INPUT_SIZE = 64 * 1024

    /** Cheap RIFF/WAVE sniff so imported non-WAV uploads pass through untouched. */
    fun isWav(file: File): Boolean {
        if (file.length() < 12) return false
        RandomAccessFile(file, "r").use { raf ->
            val head = ByteArray(12)
            if (raf.read(head) < 12) return false
            return String(head, 0, 4, Charsets.US_ASCII) == "RIFF" &&
                String(head, 8, 4, Charsets.US_ASCII) == "WAVE"
        }
    }

    /** Walks RIFF chunks to read format + locate PCM data. Tolerates extra chunks. */
    fun readWavFormat(file: File): WavFormat {
        RandomAccessFile(file, "r").use { raf ->
            val riff = ByteArray(4).also { raf.readFully(it) }
            require(String(riff, Charsets.US_ASCII) == "RIFF") { "Not a RIFF file" }
            raf.skipBytes(4) // overall RIFF size
            val wave = ByteArray(4).also { raf.readFully(it) }
            require(String(wave, Charsets.US_ASCII) == "WAVE") { "Not a WAVE file" }

            var sampleRate = 0
            var channels = 0
            var haveFmt = false
            val tag = ByteArray(4)

            while (true) {
                if (raf.read(tag) < 4) throw IllegalArgumentException("WAV missing 'data' chunk")
                val size = readUInt32LE(raf)
                when (String(tag, Charsets.US_ASCII)) {
                    "fmt " -> {
                        val fmt = ByteArray(16).also { raf.readFully(it) }
                        val bb = ByteBuffer.wrap(fmt).order(ByteOrder.LITTLE_ENDIAN)
                        bb.short // audio format (PCM = 1)
                        channels = bb.short.toInt() and 0xFFFF
                        sampleRate = bb.int
                        haveFmt = true
                        val skip = size - 16 + (size and 1L) // remainder + pad byte
                        if (skip > 0) raf.seek(raf.filePointer + skip)
                    }
                    "data" -> {
                        require(haveFmt) { "WAV 'data' chunk preceded 'fmt '" }
                        require(channels >= 1) { "WAV has no channels" }
                        return WavFormat(sampleRate, channels, raf.filePointer, size)
                    }
                    else -> raf.seek(raf.filePointer + size + (size and 1L))
                }
            }
        }
    }

    /**
     * Transcodes [source] (PCM WAV) to AAC-LC in an `.m4a` at [dest], preserving
     * sample rate + channel count. Streams the source so multi-hour recordings
     * don't load fully into memory. Overwrites [dest] if present.
     */
    fun compressWavToM4a(source: File, dest: File, bitRate: Int = 48_000): CompressedAudio {
        val wav = readWavFormat(source)
        require(wav.dataLength > 0) { "WAV contains no PCM data" }
        if (dest.exists()) dest.delete()

        val mediaFormat = MediaFormat.createAudioFormat(
            MediaFormat.MIMETYPE_AUDIO_AAC, wav.sampleRate, wav.channels,
        ).apply {
            setInteger(MediaFormat.KEY_AAC_PROFILE, MediaCodecInfo.CodecProfileLevel.AACObjectLC)
            setInteger(MediaFormat.KEY_BIT_RATE, bitRate)
            setInteger(MediaFormat.KEY_MAX_INPUT_SIZE, MAX_INPUT_SIZE)
        }

        val codec = MediaCodec.createEncoderByType(MediaFormat.MIMETYPE_AUDIO_AAC)
        val muxer = MediaMuxer(dest.absolutePath, MediaMuxer.OutputFormat.MUXER_OUTPUT_MPEG_4)
        val bytesPerFrame = 2 * wav.channels

        try {
            codec.configure(mediaFormat, null, null, MediaCodec.CONFIGURE_FLAG_ENCODE)
            codec.start()

            RandomAccessFile(source, "r").use { raf ->
                raf.seek(wav.dataOffset)
                var remaining = wav.dataLength
                var framesRead = 0L
                var trackIndex = -1
                var muxerStarted = false
                var inputDone = false
                val info = MediaCodec.BufferInfo()

                while (true) {
                    if (!inputDone) {
                        val inIndex = codec.dequeueInputBuffer(TIMEOUT_US)
                        if (inIndex >= 0) {
                            val inBuf = codec.getInputBuffer(inIndex)!!.apply { clear() }
                            val toRead = minOf(inBuf.capacity().toLong(), remaining).toInt()
                            var filled = 0
                            if (toRead > 0) {
                                val chunk = ByteArray(toRead)
                                val n = raf.read(chunk, 0, toRead)
                                if (n > 0) {
                                    inBuf.put(chunk, 0, n)
                                    filled = n
                                }
                            }
                            val ptsUs = framesRead * 1_000_000L / wav.sampleRate
                            if (filled <= 0) {
                                codec.queueInputBuffer(inIndex, 0, 0, ptsUs, MediaCodec.BUFFER_FLAG_END_OF_STREAM)
                                inputDone = true
                            } else {
                                codec.queueInputBuffer(inIndex, 0, filled, ptsUs, 0)
                                remaining -= filled
                                framesRead += (filled / bytesPerFrame).toLong()
                            }
                        }
                    }

                    when (val outIndex = codec.dequeueOutputBuffer(info, TIMEOUT_US)) {
                        MediaCodec.INFO_OUTPUT_FORMAT_CHANGED -> {
                            trackIndex = muxer.addTrack(codec.outputFormat)
                            muxer.start()
                            muxerStarted = true
                        }
                        MediaCodec.INFO_TRY_AGAIN_LATER -> Unit
                        else -> if (outIndex >= 0) {
                            val outBuf = codec.getOutputBuffer(outIndex)!!
                            if (info.flags and MediaCodec.BUFFER_FLAG_CODEC_CONFIG != 0) {
                                info.size = 0 // codec config is consumed by addTrack, not muxed
                            }
                            if (info.size > 0 && muxerStarted) {
                                outBuf.position(info.offset)
                                outBuf.limit(info.offset + info.size)
                                muxer.writeSampleData(trackIndex, outBuf, info)
                            }
                            codec.releaseOutputBuffer(outIndex, false)
                            if (info.flags and MediaCodec.BUFFER_FLAG_END_OF_STREAM != 0) break
                        }
                    }
                }
            }
        } finally {
            runCatching { codec.stop() }
            codec.release()
            runCatching { muxer.stop() }
            muxer.release()
        }

        val durationSeconds = wav.dataLength / bytesPerFrame / wav.sampleRate
        return CompressedAudio(dest, dest.length(), durationSeconds)
    }

    private fun readUInt32LE(raf: RandomAccessFile): Long {
        val b = ByteArray(4).also { raf.readFully(it) }
        return (b[0].toLong() and 0xFF) or
            ((b[1].toLong() and 0xFF) shl 8) or
            ((b[2].toLong() and 0xFF) shl 16) or
            ((b[3].toLong() and 0xFF) shl 24)
    }
}
